fsspec/implementations/reference.py

```diff
@@ -53,32 +51,6 @@ def _protocol_groups(paths, references):
     return out
 
 
-class RefsValuesView(collections.abc.ValuesView):
```

Comment:
I personally think it would be very worthwhile to include these changes because they make dramatic differences in performance when iterating over the entire reference set sequentially. I'll explain why in more detail in a bit, but just to give you an idea of the significance of this it takes nearly 40 minutes for me to iterate over the entire NWM reference set (which is a lot slower than the original JSON, even after including the load time for that). through the lazy ref mapper with the default implementation of `values()` and `getitems()`. When overriding with these implementations, it takes just 2 minutes, so nearly a 20x improvement.

The main reason why is the default implementations of `values()` and `getitems()` call `__getitem__()` at every iteration, and this has a ton of overhead compared to retrieving a key from a normal dictionary. In particular it has to:

- Convert the string chunk key to a 1D integer index
- Figure out which parquet file to load from and from which row
- Check whether each item is a raw or url
- Additionally if the urls are categorical indexing into it is also a lot slower.

These are all fine for random access since the amount of time it will take to load the actual chunk will be much greater than the time it takes to load each reference, but for situations where we just need to process all references and not actually load the data (perhaps in combine?), this adds up really fast.

So basically, this custom implementation takes advantage of the fact that we know the correct order of the chunk keys in advance, we can just iterate them separately rather than explicitly calling `__getitem__()` so we can easily eliminate the first two bottlenecks from having to process each key. And since each iteration is done over each parquet file, we can greatly reduce the number of times we need to check for whether the reference is a raw or url since this can be checked once per file if there are less than 4 columns. Finally categorical URL arrays are fully expanded so they can be indexed into much faster. In this way the four bottlenecks I mentioned above are eliminated.

Reply:
You make some very good points! I guess this would be a critical failure when combining if it takes so long. Would you like me to undo these changes, or will you?

Reply:
Great! I can go ahead and do this.